feat(netlify): support included/excluded files (#325)
withastro/adapters: .changeset/itchy-donuts-love.md

---
'@astrojs/netlify': minor
---

Adds "includedFiles" and "excludedFiles" options. These allow extra files to be deployed in the SSR function bundle.

> When an Astro site using `server` or `hybrid` rendering is deployed to Netlify, the generated functions trace the server dependencies and include any that may be needed in SSR. However sometimes you may want to include extra files that are not detected as dependencies, such as files that are loaded using `fs` functions. Also, you may sometimes want to specifically exclude dependencies that are bundled automatically. For example, you may have a Node module that includes a large binary. The `"includedFiles"` and `"excludedFiles"` options allow you to do these. Both options support glob patterns, so you can include/exclude multiple files at once.

Reviewer: ```suggestion
When an Astro site using `server` or `hybrid` rendering is deployed to Netlify, the generated functions trace the server dependencies and include any that may be needed in SSR. However sometimes you may want to include extra files that are not detected as dependencies, such as files that are loaded using `fs` functions. Also, you may sometimes want to specifically exclude dependencies that are bundled automatically. For example, you may have a Node module that includes a large binary.

The `includedFiles` and `excludedFiles` options allow you specify these inclusions and exclusions as an array of file paths relative to the project root. Both options support glob patterns, so you can include/exclude multiple files at once.
```
Just thought this paragraph was getting a *little* long, so maybe breaking it up to separate where the use case is described vs the options themselves?

> The paths are relative to the site root. If you are loading them using filesystem funcitons, make sure you resolve paths relative to the site root, and not relative to the cxource file. At runtime the compiled file will be in a different location, so paths that are relative to the file will not work. You should instead resolve the path using `path.resolve()` or `process.cwd()`, which will give you the site root.

Reviewer 2: I think it's worth showing a little snippet of do/don't

Author: Good idea

```js
import netlify from '@astrojs/netlify';
import { defineConfig } from 'astro/config';

export default defineConfig({
  output: 'server',
  adapter: netlify({
    includedFiles: ['files/**/*.csv', 'files/include-this.txt'],

>     excludedFiles: ['files/subdirectory/not-this.csv'],

Reviewer: Not blocking, but just throwing it out there: since most people's projects are maybe not likely to have a folder actually called `files/`, I'm whether there's a starting directory that represents something someone *would* have in their project, and these inclusions/exclusions could be shown for a specific reason?

(I snagged from the code to know that these are relative to the project root, because I wasn't immediately sure what or where `files/` might be, since it's not a common example we use! But, with that added, the example is probably easy enough to grok and I'll let you decide.)

For example, I think it *could* be an idea to show excluding something from node modules, if that's something people are likely to use this for.

The `/subdirectory/` and `include-this`/`not-this` style names are great! (Even something bordering on the style of `not-this-way-huge-binary` e.g., if you're feeling cheeky!)

Author: Great suggestions! I changed it to:

```js
    includedFiles: ['src/address-data/**/*.csv', 'src/include-this.txt'],
    excludedFiles: ['node_modules/chonky-module/not-this-massive-file.mp4']
```

---
'@astrojs/netlify': minor
---

> Adds "includedFiles" and "excludedFiles" options. These allow extra files to be deployed in the SSR function bundle.

Reviewer: ```suggestion
Adds `includedFiles` and `excludedFiles` configuration options. These allow extra files to be deployed in the SSR function bundle.
```

> When an Astro site using `server` or `hybrid` rendering is deployed to Netlify, the generated functions trace the server dependencies and include any that may be needed in SSR. However sometimes you may want to include extra files that are not detected as dependencies, such as files that are loaded using `fs` functions. Also, you may sometimes want to specifically exclude dependencies that are bundled automatically. For example, you may have a Node module that includes a large binary.

Reviewer: ```suggestion
When an Astro site using `server` or `hybrid` rendering is deployed to Netlify, the generated functions trace the server dependencies and include any that may be needed in SSR. However, sometimes you may want to include extra files that are not detected as dependencies, such as files that are loaded using `fs` functions. Also, you may sometimes want to specifically exclude dependencies that are bundled automatically. For example, you may have a Node module that includes a large binary.
```

> The `includedFiles` and `excludedFiles` options allow you specify these inclusions and exclusions as an array of file paths relative to the project root. Both options support glob patterns, so you can include/exclude multiple files at once.
> 
> The paths are relative to the site root. If you are loading them using filesystem functions, make sure you resolve paths relative to the site root, and not relative to the source file. At runtime the compiled file will be in a different location, so paths that are relative to the file will not work. You should instead resolve the path using `path.resolve()` or `process.cwd()`, which will give you the site root.

Reviewer: ```suggestion
The `includedFiles` and `excludedFiles` options allow you specify these inclusions and exclusions as an array of file paths relative to the site root. Both options support glob patterns, so you can include/exclude multiple files at once.

If you are specifying files using filesystem functions, resolve the path using `path.resolve()` or `process.cwd()`, which will give you the site root. At runtime, compiled source files will be in a different location and you cannot rely relative file paths.
```

Just noting the above seemed a little repetative (array of paths relative to the site root.. the paths are relative to the site root... resolve paths relative to the site root...) :smile: 

Something like the above maybe conveys the info more compactly, and doesn't make people read more than necessary to get the info they need? (Will note that as a *changeset*, this doesn't have to be full, cover-everything, documentation either! We can spell things out more clearly and exhuastively in docs itself if you think that's necessary!)

Do you think something like this does a good enough "changelog" job?

> The paths are relative to the site root. If you are loading them using filesystem funcitons, make sure you resolve paths relative to the site root, and not relative to the cxource file. At runtime the compiled file will be in a different location, so paths that are relative to the file will not work. You should instead resolve the path using `path.resolve()` or `process.cwd()`, which will give you the site root.

Reviewer 3: ```suggestion
The paths are relative to the site root. If you are loading them using filesystem functions, make sure you resolve paths relative to the site root, and not relative to the source file. At runtime the compiled file will be in a different location, so paths that are relative to the file will not work. You should instead resolve the path using `path.resolve()` or `process.cwd()`, which will give you the site root.
```
